### PR TITLE
rpc : make RPC servers come first in the device list

### PR DESCRIPTION
**Pull request by rgerganov**
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

This patch implements an idea suggested by @slaren which is to make RPC servers come first in the device list. When the last device is a local one, we don't have to transfer the logits over the network and this have significant impact on the performance.

Here are the results for my laptop (NVIDIA T1200) offloading to an RPC server with Tesla T4 on 100Mb/s network:

| GPU                     | Model           |   Model Size [GiB] |   Num. of Par. | Test   |   t/s master |   t/s rpc-first |   Speedup |
|:------------------------|:----------------|-------------------:|---------------:|:-------|-------------:|----------------:|----------:|
| NVIDIA T1200 Laptop GPU | llama 1B Q8_0   |               1.09 |     1100048384 | pp512  |      1742.68 |         1987.90 |      1.14 |
| NVIDIA T1200 Laptop GPU | llama 1B Q8_0   |               1.09 |     1100048384 | tg128  |        67.98 |           82.49 |      1.21 |
| NVIDIA T1200 Laptop GPU | llama 8B Q4_K_M |               4.58 |     8030261248 | pp512  |       356.96 |          465.93 |      1.31 |
| NVIDIA T1200 Laptop GPU | llama 8B Q4_K_M |               4.58 |     8030261248 | tg128  |        27.36 |           29.95 |      1.09 |

The amount of network traffic sent from the RPC server to the main host is from 4x to 15x less with this patch.

### CODE REVIEW

File: src/llama.cpp [hunk 1]
```diff
@@ -3346,37 +3346,37 @@ static size_t llama_get_device_count(const llama_model & model) {
 static ggml_backend_buffer_type_t llama_default_buffer_type_offload(const llama_model & model, int gpu) {
     ggml_backend_buffer_type_t buft = nullptr;
 
-#if defined(GGML_USE_RPC)
-    int dev_count = (int)llama_get_device_count(model);
     int rpc_count = (int)model.rpc_servers.size();
-    if (gpu >= dev_count - rpc_count) {
-        const char * endpoint = model.rpc_servers[gpu - dev_count + rpc_count].c_str();
+    int local_gpu = gpu - rpc_count;
```
Comment (slaren) [inline src/llama.cpp:3350]: I don't think this will work correctly if a list of RPC servers is given in a build without the RPC backend (they should be ignored). The device ids should be from 0 to `llama_get_device_count() - 1`.

Comment (rgerganov) [reply src/llama.cpp:3350]: I think giving a list of RPC servers in non-RPC build should produce an error, i.e. the `--rpc` command line option must not be available.

Comment (slaren) [reply src/llama.cpp:3350]: Fixing that in the examples is good, but it still needs to be handled properly for 3rd party applications that use the llama.cpp API directly. `llama_model::rpc_servers` should probably only exist in builds with the RPC backend.

Comment (rgerganov) [reply src/llama.cpp:3354]: I have added `ifdefs` for the `rpc_count`. If we want to remove the existence of `llama_model::rpc_server`, we will need more `ifdefs`.

**Review by slaren — APPROVED** [review]

**Review by ggerganov — APPROVED** [review]

**Comment by awatuna** [comment]
I think this broke the Windows cuda llama-server's rpc.

The Windows cuda llama-server used to ignore --rpc.

However if you build it with GGML_RPC set then the Windows cuda llama-server, eg b3649 with #9184 patch, works utilizing both local Windows gpu and remote Linux gpu.

Now build with GGML_RPC set, the Windows cuda llama-server ignores --rpc again.

**Comment by rgerganov** [comment]
> Now build with GGML_RPC set, the Windows cuda llama-server ignores --rpc again.

Hm, I don't have a windows machine atm but I cannot reproduce this on Linux. What do you mean the `--rpc` option is ignored? Can you start `rpc-server` on localhost and paste the output of:

```
llama-server -m <some_model.gguf> --rpc localhost:50052 -ngl 99
```

**Comment by awatuna** [comment]
The official Windows cuda build
https://github.com/ggerganov/llama.cpp/releases/download/b3669/llama-b3669-bin-win-cuda-cu12.2.0-x64.zip

It does not build rpc-server, and the llama-server prints help when --rpc is set.
No rpc at all.

The merged patch from another collaborator that enables rpc by default,
only enabled rpc for Windows builds, not Windows cuda builds.
Windows builds and Windows cuda builds are in different parts of the CI workflow, the patch missed it.
So I made another patch, #9184 to enable it.

The resulting b3649 with GGML_RPC set for Windows cuda build
https://github.com/awatuna/llama.cpp/releases/tag/b3655

llama-server rpc works

llm_load_tensors: ggml ctx size =    0.68 MiB
llm_load_tensors: offloading 46 repeating layers to GPU
llm_load_tensors: offloading non-repeating layers to GPU
llm_load_tensors: offloaded 47/47 layers to GPU
llm_load_tensors: RPC[192.168.200.246:50052] buffer size =  9228.81 MiB
llm_load_tensors:        CPU buffer size =  1195.31 MiB
llm_load_tensors:      CUDA0 buffer size = 18362.25 MiB

However, b3664 with GGML_RPC set for Windows cuda build
https://github.com/awatuna/llama.cpp/releases/tag/b3671

llama-server accepts --rpc, but it didn't even try to connect to rpc server.

llm_load_tensors: ggml ctx size =    0.68 MiB
llama_model_load: error loading model: unable to allocate backend buffer
llama_load_model_from_file: failed to load model

**Comment by awatuna** [comment]
If the rpc server is down, win cuda b3649 with rpc will spam

Failed to connect to 192.168.200.246:50052

win cuda b3664 with rpc just ignores the --rpc, doesn't try to connect at all.
